core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -233,10 +233,16 @@ private[log] class ProducerAppendInfo(val topicPartition: TopicPartition,
 
       // If there is no current producer epoch (possibly because all producer records have been deleted due to
       // retention or the DeleteRecords API) accept writes with any sequence number
-      if (!(currentEntry.producerEpoch == RecordBatch.NO_PRODUCER_EPOCH || inSequence(currentLastSeq, appendFirstSeq))) {
+      if (currentEntry.producerEpoch != RecordBatch.NO_PRODUCER_EPOCH && !inSequence(currentLastSeq, appendFirstSeq)) {
         throw new OutOfOrderSequenceException(s"Out of order sequence number for producer $producerId at " +
-          s"offset $offset in partition $topicPartition: $appendFirstSeq (incoming seq. number), " +
-          s"$currentLastSeq (current end sequence number)")
+          s"offset $offset in partition $topicPartition: producer request epoch $producerEpoch, " +
+          s"current entry epoch ${currentEntry.producerEpoch}, " +
+          s"updating entry epoch ${updatedEntry.producerEpoch}, " +
+          s"append first seq $appendFirstSeq, " +
+          s"updating entry last seq ${updatedEntry.lastSeq}, " +
+          s"updating batch empty ${updatedEntry.isEmpty}" +
```

Comment:
nit: updating entry last seq will be `-1` if `updatedEntry.isEmpty`. Is this line still necessary? Or you think it's possible this value might be changed by other thread?